Ikemen go 99.0 stable version the true type font is not read in the system.sff

### Describe the bug

the error is posted on the `freetype: invalid TrueType format: bad kern table length
goroutine 1 [running, locked to thread]:
github.com/yuin/gopher-lua.(*LState).PCall.func1()
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/state.go:1988 +0x212
panic({0x7ff7a0f6a380, 0x7ff7a108d708})
	C:/hostedtoolcache/windows/go/1.20.10/x64/src/runtime/panic.go:884 +0x213
main.LoadFntTtf(0xc000389950, {0xc007c0e6c0?, 0x7ff7a0fd53c8?}, {0xc006f40a4b?, 0x7ff7a108d410?}, 0xffffffff)
	D:/a/Ikemen-GO/Ikemen-GO/src/util_desktop.go:49 +0x245
main.loadDefInfo(0xc000389950, {0xc007c0e6c0, 0x1f}, 0x7ff7a0fbf8e0?, 0x7490e08?)
	D:/a/Ikemen-GO/Ikemen-GO/src/font.go:308 +0x428
main.loadFntV2({0xc007c0e6c0, 0x1f}, 0xa0fd518c?)
	D:/a/Ikemen-GO/Ikemen-GO/src/font.go:271 +0x209
main.loadFnt({0xc007c0e6c0, 0x1f}, 0x7491048?)
	D:/a/Ikemen-GO/Ikemen-GO/src/font.go:52 +0x65
main.systemScriptInit.func56(0xc0001186e0)
	D:/a/Ikemen-GO/Ikemen-GO/src/script.go:844 +0x128
github.com/yuin/gopher-lua.callGFunction(0xc0001186e0, 0x0)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/vm.go:202 +0x37
github.com/yuin/gopher-lua.init.3.func26(0xc0001186e0, 0x60100a04?, 0xc0002050a0?)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/vm.go:821 +0x379
github.com/yuin/gopher-lua.mainLoop(0xc0001186e0, 0xc00816e840?)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/vm.go:31 +0xfa
github.com/yuin/gopher-lua.(*LState).callR(0xc0001186e0, 0x1, 0x1, 0xc000018ea0?)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/state.go:1211 +0x1b4
github.com/yuin/gopher-lua.(*LState).Call(...)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/state.go:1967
github.com/yuin/gopher-lua.loRequire(0xc0001186e0)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/baselib.go:564 +0x6d3
github.com/yuin/gopher-lua.callGFunction(0xc0001186e0, 0x0)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/vm.go:202 +0x37
github.com/yuin/gopher-lua.init.3.func26(0xc0001186e0, 0x34780600?, 0xc000205000?)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/vm.go:821 +0x379
github.com/yuin/gopher-lua.mainLoop(0xc0001186e0, 0xc000790740?)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/vm.go:31 +0xfa
github.com/yuin/gopher-lua.(*LState).callR(0xc0001186e0, 0x0, 0xffffffffffffffff, 0xc00000a7e8?)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/state.go:1211 +0x1b4
github.com/yuin/gopher-lua.(*LState).Call(...)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/state.go:1967
github.com/yuin/gopher-lua.(*LState).PCall(0xc0001186e0, 0x0, 0xc000790740?, 0x0)
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/state.go:2030 +0x145
github.com/yuin/gopher-lua.(*LState).DoFile(0xc0001186e0, {0xc000018918?, 0x0?})
	C:/Users/runneradmin/go/pkg/mod/github.com/yuin/gopher-lua@v0.0.0-20220504180219-658193537a64/auxlib.go:396 +0x6b
main.main()
	D:/a/Ikemen-GO/Ikemen-GO/src/main.go:90 +0x218

stack traceback:
	[G]: in function 'fontNew'
	external/script/main.lua:471: in function <external/script/main.lua:445>
	(tailcall): ?
	.\external\script\start.lua:1925: in function <.\external\script\start.lua:0>
	[G]: in function 'require'
	external/script/main.lua:2283: in main chunk
	[G]: ?
`
here the file attachment.
https://drive.google.com/file/d/188I0GoixIQ1LIcbyTm99_1fEkRpcZNAb/view?usp=sharing

### To Reproduce

just an issue with the fonts.

### Expected behavior

its crashing because. its will not ready the true type font

### Screenshots / Video

![Untitled](https://github.com/ikemen-engine/Ikemen-GO/assets/35203194/b5b1a97f-a06f-4caa-9725-37f28221d427)
this the cuase fo the issue

### Engine Version (or source code date)

v0.99.0

### Operating system

Windows

### Extra context or search terms

![Untitled](https://github.com/ikemen-engine/Ikemen-GO/assets/35203194/b1163bc8-f94f-44a2-b5e1-78f65e78a5c0)

```
Last login: Thu Jul 11 14:56:17 on ttys003
/Users/shinriyo/Downloads/Ikemen_GO-v0.99.0-macos/Ikemen_GO_MacOS ; exit;
(base) ~ % /Users/shinriyo/Downloads/Ikemen_GO-v0.99.0-macos/Ikemen_GO_MacOS ; exit;
panic: Main lua file "external/script/main.lua" error.
open external/script/main.lua: no such file or directory

goroutine 1 [running, locked to thread]:
main.main()
	/Users/runner/work/Ikemen-GO/Ikemen-GO/src/main.go:80 +0x529

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]


```

--- a/src/font.go
+++ b/src/font.go
@@ -100,6 +100,43 @@ func newFnt() *Fnt {
 	}
 }
 
+// workaround for https://github.com/golang/freetype/issues/8
+func stripKernTable(ttf []byte) []byte {
+	if len(ttf) < 12 {
+		return ttf
+	}
+
+	offset := 0
+	if string(ttf[:4]) == "ttcf" {
+		if len(ttf) < 16 {
+			return ttf
+		}
+		offset = int(binary.BigEndian.Uint32(ttf[12:16]))
+		if offset < 0 || len(ttf)-offset < 12 {
+			return ttf
+		}
+	}
+
+	numTables := int(binary.BigEndian.Uint16(ttf[offset+4 : offset+6]))
+	tableDir := offset + 12
+	if len(ttf) < tableDir+16*numTables {
+		return ttf
+	}
+
+	out := append([]byte(nil), ttf...)
+	for i := 0; i < numTables; i++ {
+		p := tableDir + 16*i
+		if string(out[p:p+4]) == "kern" {
+			// Keep the table directory entry, but make the table empty.
+			// golang/freetype will then ignore kerning instead of failing Parse.
+			binary.BigEndian.PutUint32(out[p+12:p+16], 0)
+			break
+		}
+	}
+
+	return out
+}
+
 func loadFnt(filename string, height int32) (*Fnt, error) {
 	if HasExtension(filename, ".fnt") {
 		return loadFntV1(filename)

--- a/src/font_gl33.go
+++ b/src/font_gl33.go
@@ -354,9 +354,18 @@ func (f *Font_GL33) GenerateGlyphs(low, high rune) error {
 	for ch := low; ch <= high; ch++ {
 		char := new(character)
 
+		drawGlyph := true
 		gBnd, gAdv, ok := ttfFace.GlyphBounds(ch)
-		if ok != true {
-			return fmt.Errorf("ttf face glyphBounds error")
+		if !ok {
+			// Some fonts do not provide bounds for every requested rune. This is common for control chars.
+			// Keep loading the font and create an invisible placeholder so future lookups do not retry this glyph forever.
+			drawGlyph = false
+			if adv, advOK := ttfFace.GlyphAdvance(ch); advOK {
+				gAdv = adv
+			} else {
+				gAdv = 0
+			}
+			gBnd = f.ttf.Bounds(fixed.Int26_6(f.scale))
 		}
 
 		gh := int32((gBnd.Max.Y - gBnd.Min.Y) >> 6)
@@ -408,9 +417,22 @@ func (f *Font_GL33) GenerateGlyphs(low, high rune) error {
 		c.SetClip(rgba.Bounds())
 		c.SetDst(rgba)
 		c.SetSrc(fg)
-		_, err := c.DrawString(string(ch), pt)
-		if err != nil {
-			return err
+		if drawGlyph {
+			if _, err := c.DrawString(string(ch), pt); err != nil {
+				// Some fonts have broken hinting bytecode. Keep full hinting for normal fonts.
+				// Retry glyph without hinting instead of failing the whole font load.
+				c2 := freetype.NewContext()
+				c2.SetDPI(72)
+				c2.SetFont(f.ttf)
+				c2.SetFontSize(float64(f.scale))
+				c2.SetHinting(font.HintingNone)
+				c2.SetClip(rgba.Bounds())
+				c2.SetDst(rgba)
+				c2.SetSrc(fg)
+				if _, err := c2.DrawString(string(ch), pt); err != nil {
+					return err
+				}
+			}
 		}
 
 		var uv [4]float32

--- a/src/font_gl33.go
+++ b/src/font_gl33.go
@@ -488,6 +488,9 @@ func (r *FontRenderer_GL33) LoadTrueTypeFont(reader io.Reader, scale int32, low,
 
 	// Read the truetype font.
 	ttf, err := truetype.Parse(data)
+	if err != nil && err.Error() == "freetype: invalid TrueType format: bad kern table length" {
+		ttf, err = truetype.Parse(stripKernTable(data))
+	}
 	if err != nil {
 		return nil, err
 	}

--- a/src/font_gles32.go
+++ b/src/font_gles32.go
@@ -354,9 +354,18 @@ func (f *Font_GLES32) GenerateGlyphs(low, high rune) error {
 	for ch := low; ch <= high; ch++ {
 		char := new(character)
 
+		drawGlyph := true
 		gBnd, gAdv, ok := ttfFace.GlyphBounds(ch)
-		if ok != true {
-			return fmt.Errorf("ttf face glyphBounds error")
+		if !ok {
+			// Some fonts do not provide bounds for every requested rune. This is common for control chars.
+			// Keep loading the font and create an invisible placeholder so future lookups do not retry this glyph forever.
+			drawGlyph = false
+			if adv, advOK := ttfFace.GlyphAdvance(ch); advOK {
+				gAdv = adv
+			} else {
+				gAdv = 0
+			}
+			gBnd = f.ttf.Bounds(fixed.Int26_6(f.scale))
 		}
 
 		gh := int32((gBnd.Max.Y - gBnd.Min.Y) >> 6)
@@ -408,10 +417,23 @@ func (f *Font_GLES32) GenerateGlyphs(low, high rune) error {
 		c.SetClip(rgba.Bounds())
 		c.SetDst(rgba)
 		c.SetSrc(fg)
-		_, err := c.DrawString(string(ch), pt)
-		if err != nil {
-			Logcat(fmt.Sprintf("GLES: ERROR DRAWING STRING: %v", err.Error()))
-			return err
+		if drawGlyph {
+			if _, err := c.DrawString(string(ch), pt); err != nil {
+				// Some fonts have broken hinting bytecode. Keep full hinting for normal fonts.
+				// Retry glyph without hinting instead of failing the whole font load.
+				c2 := freetype.NewContext()
+				c2.SetDPI(72)
+				c2.SetFont(f.ttf)
+				c2.SetFontSize(float64(f.scale))
+				c2.SetHinting(font.HintingNone)
+				c2.SetClip(rgba.Bounds())
+				c2.SetDst(rgba)
+				c2.SetSrc(fg)
+				if _, err := c2.DrawString(string(ch), pt); err != nil {
+					Logcat(fmt.Sprintf("GLES: ERROR DRAWING STRING: %v", err.Error()))
+					return err
+				}
+			}
 		}
 		// Logcat(fmt.Sprintf("Char: %c | Box: %dx%d | Dot: %v | Bounds: %v\n", ch, gw, gh, pt, gBnd))
 

--- a/src/font_gles32.go
+++ b/src/font_gles32.go
@@ -495,6 +495,9 @@ func (r *FontRenderer_GLES32) LoadTrueTypeFont(reader io.Reader, scale int32, lo
 
 	// Read the truetype font.
 	ttf, err := truetype.Parse(data)
+	if err != nil && err.Error() == "freetype: invalid TrueType format: bad kern table length" {
+		ttf, err = truetype.Parse(stripKernTable(data))
+	}
 	if err != nil {
 		return nil, err
 	}

--- a/src/font_vk.go
+++ b/src/font_vk.go
@@ -374,9 +374,18 @@ func (f *Font_VK) GenerateGlyphs(low, high rune) error {
 	for ch := low; ch <= high; ch++ {
 		char := new(character)
 
+		drawGlyph := true
 		gBnd, gAdv, ok := ttfFace.GlyphBounds(ch)
-		if ok != true {
-			return fmt.Errorf("ttf face glyphBounds error")
+		if !ok {
+			// Some fonts do not provide bounds for every requested rune. This is common for control chars.
+			// Keep loading the font and create an invisible placeholder so future lookups do not retry this glyph forever.
+			drawGlyph = false
+			if adv, advOK := ttfFace.GlyphAdvance(ch); advOK {
+				gAdv = adv
+			} else {
+				gAdv = 0
+			}
+			gBnd = f.ttf.Bounds(fixed.Int26_6(f.scale))
 		}
 
 		gh := int32((gBnd.Max.Y - gBnd.Min.Y) >> 6)
@@ -430,9 +439,22 @@ func (f *Font_VK) GenerateGlyphs(low, high rune) error {
 		c.SetClip(rgba.Bounds())
 		c.SetDst(rgba)
 		c.SetSrc(fg)
-		_, err := c.DrawString(string(ch), pt)
-		if err != nil {
-			return err
+		if drawGlyph {
+			if _, err := c.DrawString(string(ch), pt); err != nil {
+				// Some fonts have broken hinting bytecode. Keep full hinting for normal fonts.
+				// Retry glyph without hinting instead of failing the whole font load.
+				c2 := freetype.NewContext()
+				c2.SetDPI(72)
+				c2.SetFont(f.ttf)
+				c2.SetFontSize(float64(f.scale))
+				c2.SetHinting(font.HintingNone)
+				c2.SetClip(rgba.Bounds())
+				c2.SetDst(rgba)
+				c2.SetSrc(fg)
+				if _, err := c2.DrawString(string(ch), pt); err != nil {
+					return err
+				}
+			}
 		}
 
 		// Generate texture

--- a/src/font_vk.go
+++ b/src/font_vk.go
@@ -526,6 +526,9 @@ func (r *FontRenderer_VK) LoadTrueTypeFont(reader io.Reader, scale int32, low, h
 
 	// Read the truetype font.
 	ttf, err := truetype.Parse(data)
+	if err != nil && err.Error() == "freetype: invalid TrueType format: bad kern table length" {
+		ttf, err = truetype.Parse(stripKernTable(data))
+	}
 	if err != nil {
 		return nil, err
 	}

--- a/src/util_desktop.go
+++ b/src/util_desktop.go
@@ -46,7 +46,6 @@ func LoadFntTtf(f *Fnt, fontfile string, filename string, height int32) {
 	}
 	ttf, err := gfxFont.LoadFont(fileDir, height, int(sys.gameWidth), int(sys.gameHeight))
 	if err != nil {
-		panic(err)
 		panic(fmt.Errorf("failed to load ttf font %v: %w", fileDir, err))
 	}
 	f.ttf = ttf.(Font)

